firebase.go

```diff
@@ -0,0 +1,147 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"net/http"
+	"os"
+
+	"cloud.google.com/go/firestore"
+	firebase "firebase.google.com/go"
+	"firebase.google.com/go/auth"
+	"google.golang.org/api/option"
+)
+
+var (
+	ctx        context.Context
+	authClient *auth.Client
+	fsClient   *firestore.Client
+)
+
+func handleRequest(w http.ResponseWriter, r *http.Request) {
+	idToken := r.FormValue("idToken")
+	init_firebase()
+	user, err := authenticateUser(ctx, authClient, idToken)
+	if err != nil {
+		log.Fatalf("authenticateUser: %v", err)
+	}
+	docRef := fsClient.Collection("users").Doc(user.UID)
+	_, err = docRef.Set(ctx, map[string]interface{}{
+		"name":  user.DisplayName,
+		"email": user.Email,
+	})
+	if err != nil {
+		log.Fatalf("docRef.Set: %v", err)
+	}
+
+	// Add user preferences to the Firestore document
+	_, err = docRef.Update(ctx, []firestore.Update{
+		{
+			Path:  "preference",
+			Value: "Preference value",
+		},
+	})
+	if err != nil {
+		log.Fatalf("docRef.Update: %v", err)
```

Comment:
fatalf will shut down the whole system on error, like the entire website, this needs to handle failures more gracefully

Reply:
un-fatalf'd